tests/deberta_v2/test_tokenization_deberta_v2.py

```diff
@@ -67,6 +67,144 @@ def test_get_vocab(self):
     def test_vocab_size(self):
         self.assertEqual(self.get_tokenizer().vocab_size, 30_000)
 
+    def test_do_lower_case(self):
+        tokenizer = DebertaV2Tokenizer(SAMPLE_VOCAB, do_lower_case=True)
+
+        sequence = " \tHeLLo!how  \n Are yoU?  "
+
+        tokens = tokenizer.tokenize(sequence)
+
+        self.assertListEqual(tokens, ["▁hello", "!", "how", "▁are", "▁you", "?"])
+
+        if self.test_rust_tokenizer:
+            rust_tokenizer = DebertaV2TokenizerFast(SAMPLE_VOCAB, do_lower_case=True)
+
+            rust_tokens = rust_tokenizer.tokenize(sequence)
+
+            self.assertListEqual(rust_tokens, ["▁hello", "!", "how", "▁are", "▁you", "?"])
+            self.assertListEqual(tokens, rust_tokens)
```

Comment:
@SaulLu thank you for your comments! yeah I agree with you, but I wanted to keep as close as possible to the format of existing tests. So if it's fine I shall modify the rest of the tests accordingly! Just a doubt with regards to `test_full_tokenizer` test - should we keep it as it is? We are only comparing the each lists once, and it seems to me introduce more verbosity if I was to assign it like this:

```
    def test_full_tokenizer(self):
        tokenizer = DebertaV2Tokenizer(SAMPLE_VOCAB, keep_accents=True)

        tokens = tokenizer.tokenize("This is a test")
        tokens_target = ["▁", "T", "his", "▁is", "▁a", "▁test"]
        self.assertListEqual(tokens, tokens_target)

        ids_target = [13, 1, 4398, 25, 21, 1289]
        self.assertListEqual(tokenizer.convert_tokens_to_ids(tokens), ids_target)

        tokens = tokenizer.tokenize("I was born in 92000, and this is falsé.")
        tokens_target = ["▁", "I", "▁was", "▁born", "▁in", "▁9", "2000", ",", "▁and", "▁this", "▁is", "▁fal", "s", "é", "."]
        self.assertListEqual(tokens, tokens_target)

        ids = tokenizer.convert_tokens_to_ids(tokens)
        ids_target = [13, 1, 23, 386, 19, 561, 3050, 15, 17, 48, 25, 8256, 18, 1, 9]
        self.assertListEqual(ids, ids_target)

        back_tokens = tokenizer.convert_ids_to_tokens(ids)
        back_tokens_target = ["▁", "<unk>", "▁was", "▁born", "▁in", "▁9", "2000", ",", "▁and", "▁this", "▁is", "▁fal", "s", "<unk>", "."]
        self.assertListEqual(back_tokens, back_tokens_target)
```

Reply:
@SaulLu Sorry for the late response, I looked into your first point - can I check if I'm going on the right path here? It appears that the difference is due to for Debertav2 - the slow tokenizer uses the backend Sentencepiece tokenizer to tokenize https://github.com/huggingface/transformers/blob/f7196f2e63b14e9fbb4ad664e71912aab3b484cf/src/transformers/models/deberta_v2/tokenization_deberta_v2.py#L403-L410

whereas the fast tokenizer seem to use the Fast Tokenizer from the tokenizers library using the converted SPM files, so the mismatch in normalization has to be due to the backend Tokenizer of the Fast version?

```
>>> from transformers import DebertaV2TokenizerFast, DebertaV2Tokenizer
>>> model_name_v2 = "microsoft/deberta-v2-xlarge"
>>> model_name_v3 = "microsoft/deberta-v3-large"
>>> tokenizer_v2_s = DebertaV2Tokenizer.from_pretrained(model_name_v2)
>>> tokenizer_v3_f = DebertaV2TokenizerFast.from_pretrained(model_name_v3)
>>> backend_tokenizer = tokenizer_v2_s._tokenizer
>>> text = u"\u0041\u0302\u0300" + " " + u"\u0041\u0302\u0300"
>>> tokenizer_v2_s.convert_ids_to_tokens(backend_tokenizer.spm.encode(text))
['▁', '[UNK]', '▁', '[UNK]']
>>> tokenizer_v2_f.convert_ids_to_tokens(tokenizer_v2_f.encode(text))
['[CLS]', '▁Â', '▁Â', '[SEP]']
```
I've also tested using some of the other tests - [French](https://github.com/huggingface/transformers/blob/8e0d3b427f4936cd0318996d4c3dbdd779d16641/tests/camembert/test_tokenization_camembert.py#L119-L125) and [Japanese](https://github.com/huggingface/transformers/blob/8e0d3b427f4936cd0318996d4c3dbdd779d16641/tests/bert_japanese/test_tokenization_bert_japanese.py#L224)

```
>>> text = "À l'instar des réseaux de neurones récurrents (RNN), les transformeurs sont conçus "
>>> tokenizer_v2_s.convert_ids_to_tokens(tokenizer_v2_s.encode(text))
['[CLS]', '▁À', '▁l', "'", 'in', 'star', '▁des', '▁ré', 's', 'eaux', '▁de', '▁neuron', 'es', '▁ré', 'current', 's', '▁(', 'R', 'NN', '),', '▁les', '▁transform', 'eurs', '▁sont', '▁con', 'çu', 's', '[SEP]']
>>> tokenizer_v2_f.convert_ids_to_tokens(tokenizer_v2_f.encode(text))
['[CLS]', '▁À', '▁l', "'", 'in', 'star', '▁des', '▁ré', 's', 'eaux', '▁de', '▁neuron', 'es', '▁ré', 'current', 's', '▁(', 'R', 'NN', '),', '▁les', '▁transform', 'eurs', '▁sont', '▁con', 'çu', 's', '[SEP]']
>>> text = "こんにちは、世界。 \nこんばんは、世界。"
>>> tokenizer_v2_f.convert_ids_to_tokens(tokenizer_v2_f.encode(text))
['[CLS]', '▁', 'こ', 'ん', 'に', 'ち', 'は', '、', '世', '界', '。', '▁', 'こ', 'ん', 'ば', 'ん', 'は', '、', '世', '界', '。', '[SEP]']
>>> tokenizer_v2_s.convert_ids_to_tokens(tokenizer_v2_s.encode(text))
['[CLS]', '▁', 'こ', 'ん', 'に', 'ち', 'は', '、', '世', '界', '。', '▁', 'こ', 'ん', 'ば', 'ん', 'は', '、', '世', '界', '。', '[SEP]']
```
Strangely the french test with the À doesn't appear to be affected here though

Reply:
Hi @mingboiz ,

> @SaulLu thank you for your comments! yeah I agree with you, but I wanted to keep as close as possible to the format of existing tests. So if it's fine I shall modify the rest of the tests accordingly! Just a doubt with regards to test_full_tokenizer test - should we keep it as it is? We are only comparing the each lists once, and it seems to me introduce more verbosity if I was to assign it like this:

Indeed for the `test_full_tokenizer` test, I don't really have a preference for one or the other. I trust you :smile: 

### Questions about Test 1
> @SaulLu Sorry for the late response, I looked into your first point - can I check if I'm going on the right path here? It appears that the difference is due to for Debertav2 - the slow tokenizer uses the backend Sentencepiece tokenizer to tokenized 

No worries! Thank you very much for spending this amount of time for this PR! I think you are on the right track. I also spent some time investigating the problem and I agree with you: I think the difference comes from a small known difference between the normalization step in sentencepiece and the way this normalization is treated in tokenizers with the `tokenizers.normalizers.Precompiled` component (see comment [here](https://github.com/huggingface/transformers/issues/15970#issuecomment-1081639998) for more details). 

Indeed, the normalization used by `"microsoft/deberta-v2-xlarge"` is `nmt_nfkc`. The sentencepiece library stores [here](https://raw.githubusercontent.com/google/sentencepiece/master/data/nmt_nfkc.tsv)  the replacements to be carried out for this normalization. In particular, we can see the line `41 302 300	1EA6	# Ầ => Ầ` which means that `u"\u0041\u0302\u0300"` will be replaced by `u"\u1ea6"`.  And we can verify that `u"\u1ea6"` is indeed encode as an `unknown` token by our tokenizer.

**TL;DR:** I think that the origin of the difference is identified and since it is a known and quite rare "bug" I think that we can say that the **investigation for test 1 is resolved.** Is that OK?

### Questions about Test 2

I haven't put a detailed message here, but I recently [merged a PR](https://github.com/huggingface/transformers/pull/16629) to at least get a warning when a user tries to convert an spm tokenizer that uses the bytefallback feature to a fast tokenizer. I think this warning would be sufficient for now if the tokenizer fast file is not pushed to one of the `"microsoft/deberta-v3-Xxx` repos.

_I'm just pinging one of the transformers maintainers about this to get a second opinion._

@LysandreJik , @sgugger , @patrickvonplaten , @patil-suraj (I don't want to take too much time, you can just read the following paragraph) for context this PR aims to add a fast version of the tokenizer for `DebertaV2`. This class can currently be used to load `"microsoft/deberta-v2-Xxx` and `"microsoft/deberta-v3-Xxx` uploaded to the hub. One of the differences between the fast and slow versions is that with our fast tokenizers we can't enable the bytefallback option (it's in the tokenizers lib roadmap but not done). This feature is used by `"microsoft/deberta-v3-Xxx`. The current PR will only push a warning to the user that this feature is missing (cf PR #16629). Are you ok with this solution? 
I want to draw your attention to the fact that if we want to add later in our [`SpmConverter` class](https://github.com/huggingface/transformers/blob/d8269eb4d55e4e43c9b885807a82cc428783a573/src/transformers/convert_slow_tokenizer.py#L420) - which is used by several tokenizers - we will have to be very careful to push for smooth breaking change as `"microsoft/deberta-v3-Xxx` is the first tokenizer - that I'm aware of - that is in this case.

Reply:
I'm good with this solution, as long as no fast tokenizer file are updated yet for checkpoints using this option (so we can add proper ones without breaking changes once the option is supported in tokenizers).